feat(middleware): Use providers as middleware

MikeRyanDev opened the pull request:
This PR tries to solve some ergonomic issues surrounding middleware factories created via DI.

A lot of middleware will need to use the injector to resolve dependencies, like my [store-saga example](https://github.com/MikeRyan52/store-saga/blob/master/lib/index.ts#L22). While using a `multi` provider for the `PRE_MIDDLEWARE` or `POST_MIDDLEWARE` token works it restricts the consumer from being able to specify the order they want middleware to be applied.

This PR changes `usePreMiddleware` and `usePostMiddleware` to accept either middleware functions or instances of the `Provider` class. As part of resolution it will use the injector to resolve the middleware providers and compose all of the middleware in the order they are specified. To further improve their utility, these middleware functions reflect back any providers passed into them.

Demo:

``` js
const log: Middleware = input$ => input$.do(item => console.log(item));

const thunk = provide(new OpaqueToken('Thunk Middleware'), {
  deps: [Dispatcher],
  useFactory(dispatcher: Dispatcher<Action>): Middleware{
    return function(all$: Observable<Action | Thunk>){
      const [thunks$, actions$] = all$.partition(t => typeof t === 'function');

      thunks$
        .map(thunk => return new Observable(observer => {
          thunk(action => observer.next(action));
        }))
        .mergeAll()
        .subscribe(dispatcher);

      return actions$;
    }
  }
});

bootstrap(App, [ provideStore(reducer), usePreMiddleware(thunk, log) ]);
```

MikeRyanDev committed:
feat(middleware): Use providers as middleware

MikeRyanDev commented:
cc @btroncone

robwormald commented:
the idea of the `provideStore` function is that it eliminates a lot of the shenanigans of dealing with Providers in general. I'd like to see a similar API here if possible - that `thunk` has a lot of mental overhead. Not sure what the right answer is yet.

btroncone commented:
Agree, but is there a good way to hide the provider shenanigans for middleware that relies on the injector? I think a lot of this abstraction will fall on initial middleware authors. Once an ecosystem of common use-cases is established most developers will not need to touch this API. 

I will continue to think on this as well...

MikeRyanDev commented:
Maybe we can export a function called `createMiddleware` that sets up an anonymous middleware provider for you? It doesn't necessarily make creating middleware factories easier, but it may reduce the mental overhead:

``` ts
function createMiddleware(deps: any[], factoryFn: (...deps: any[]) => Middleware){
  return provide(new OpaqueToken('ngrx Middleware'), {
    deps,
    useFactory: factoryFn
  });
}

const thunk = createMiddleware(
  [Dispatcher], 
  function(dispatcher: Dispatcher<Action>): Middleware{
    return function(all$: Observable<Action | Thunk>){
      const [thunks$, actions$] = all$.partition(t => typeof t === 'function');

      thunks$
        .map(thunk => return new Observable(observer => {
          thunk(action => observer.next(action));
        }))
        .mergeAll()
        .subscribe(dispatcher);

      return actions$;
    }
  }
);
```

btroncone commented:
I like this idea. It's a small change but for the less experienced developer it abstracts away three things: `provide`, `OpaqueToken` and `useFactory`. This makes the API much more approachable. :+1:

MikeRyanDev committed:
feat(Middleware): Add simple createMiddleware utility

Add utility to create anonymous providers for middleware factories. Used to create middleware that require dependencies from the injector

MikeRyanDev commented:
Updated PR to include the `createMiddleware` utility.

robwormald commented:
looks good, though does it make sense to invert the signature? 

``` js
//no deps
createMiddleware(someFn)
//some deps
createMiddleware(someFn, Dep1, Dep2);
//or?
createMiddleware(someFn, [Dep1, Dep2])
```

MikeRyanDev commented:
Yes, I agree. Especially if we ever get type metadata for functions. I'll change the signature now.

MikeRyanDev commented:
Swapped the signature. Updated example:

``` ts
const thunk = createMiddleware(function(dispatcher: Dispatcher<Action>): Middleware{
  return function(all$: Observable<Action | Thunk>){
    const [thunks$, actions$] = all$.partition(t => typeof t === 'function');

    thunks$
      .map(thunk => return new Observable(observer => {
        thunk(action => observer.next(action));
      }))
      .mergeAll()
      .subscribe(dispatcher);

    return actions$;
  }
}, [ Dispatcher ]);
```

btroncone commented:
Looks great!

laurelnaiad commented:
It would be nice if we could apply middleware after bootstrap. Would "go great" with the ability to replace the reducer in lazy-loading scenarios where you want to "layer on" a new feature to your state machine.

I'm trying to hack something together whereby I install proxies for both pre and postMiddleware at bootstrap, and then allow for "insertions" of specific middleware later. 

I don't have this working, yet, though that may be due to other issues in my code, but it looks like the below. (It's highly likely I'm doing it wrong.)

```
const preMiddlewareSubject$ = new Subject()
const postMiddlewareSubject$ = new Subject()
function applyPreMiddleware(
  preMiddleware: (action: Observable<AppAction<any>>) =>
      Observable<AppAction<any>>
)  {
  preMiddlewareSubject$.subscribe(preMiddleware)
}

// this is what I register as the premiddleware
function preMiddleware( action: Observable<AppAction<any>> ):
    Observable<AppAction<any>> {
  action.subscribe(preMiddlewareSubject$)
  return action.do(value => preMiddlewareSubject$.next(value));
}

function applyPostiddleware(
  postMiddleware: (state: Observable<AppState>) => Observable<AppState>
)  {
  postMiddlewareSubject$.subscribe(postMiddleware)
}

// this is what I register as the postmiddleware
function postMiddleware( state: Observable<AppState> ): Observable<AppState> {
  state.subscribe(postMiddlewareSubject$)
  return state.do(value => postMiddlewareSubject$.next(value));
}

```

I do have one other comment about the ng2.ts module, while we're on the subject...

I'd like to be able to derive a custom service from Store<T>... both to make it concrete by fixing T to my ambient state tree type, but also and more importantly to add some additional capabilities to the service.  Right now, it seems I'd have to re-author the whole provideStore function due to the way the injection is arranged. If there was an OpaqueToken to represent the store service, perhaps STORE, then it would be much easier to do this. :)

MikeRyanDev commented:
Re: lazy middleware, I'm not seeing a way to make that possible. Your state is the result of applying middleware to the Dispatcher (this is also why you can't use `Store` in a middleware factory).

As for providing a custom Store service, you can do it this way:

``` ts
import {provide, Injectable, Inject} from 'angular2/core';
import {Store, Dispatcher, StoreBackend, Action, INITIAL_STATE} from '@ngrx/store';

@Injectable()
class MyStore extends Store<IState>{
  constructor(
    dispatcher: Dispatcher<Action>, 
    backend: StoreBackend, 
    @Inject(INITIAL_STATE) initialState: IState
  ){
    super(dispatcher, backend, initialState);
  }
}

// Then, in bootstrap:
bootstrap(App, [
  provideStore(reducer, initialState),
  provide(Store, { useClass: MyStore })
]);
```

laurelnaiad commented:
I don't understand why lazy-loaded middleware wouldn't be possible. 

> Your state is the result of applying middleware to the Dispatcher

So if my app runs on features a, b and c for 5 minutes, then the user navigates to the route where feature d is exposed, I just want to start "observing some stuff" that I wasn't observing before (and taking actions based on my observations, etc.) 

This would be analogous to how [redux-saga's `run` function](http://yelouafi.github.io/redux-saga/docs/api/index.html#middlewarerunsaga-args) works (at least in terms of adding middleware later in the game).

laurelnaiad commented:
P.S. Thanks for pointing me in the right direction with the dependency injection. It looks very obvious when you put it that way! :)

laurelnaiad commented:
Aside from a spelling mistake ( `applyPostiddleware` ), the above allowed me to use premiddleware on a counter so that it only decrements by adding applying this as pre-middleware after bootstrap:

```
export function spy(
  action: AppAction<any>
): AppAction<any> {
  console.log('spy', action)
  if (action.type = 'INCREMENT_COUNTER') {
    action.type = 'DECREMENT_COUNTER'
  }
  return action
}
```

When I add a second middleware, after the first, which sets all action types to INCREMENT_COUNTER, all I get is a constantly incrementing counter. In other words it sorta feels like it works.

I had actually intended for this function to be dealing in observables (or at least thought that it would be passed an observable), rather than the observed value.... but I guess since I .subscribed to the observable when I applied the the middleware, I'm getting the action itself (so I changed the signature to reflect that  :))

I did want an observable, though, so I could manage the stream itself, instead of just having cracks at the values, so I guess it's back to the drawing board. subscribing my late-loaded middleware wasn't what I wanted.

laurelnaiad commented:
@MikeRyan52 -- I've at least got something that does what I was trying to ask for ....

http://plnkr.co/edit/VBrNcJ?p=preview 

The interesting part of the plnkr for our purposes is in src/Middleware.ts.

```
import { Observable } from 'rxjs/Observable'
import { Subject } from 'rxjs/Subject'
import { Subscription } from 'rxjs/Subscription'

import {
  compose, Middleware
} from '@ngrx/store'

function recomposeMiddleware(
  middleware: Middleware[],
  subjectIn$: Observable<any>,
  subjectOut$: Observable<any>,
  subscriptionIn?: Subscription,
  subscriptionOut?: Subscription
) {
  subscriptionIn && subscriptionIn.unsubscribe()
  subscriptionOut && subscriptionOut.unsubscribe()

  let composed = compose(...middleware)

  let mySubjectIn$ = new Subject();
  let subsIn = subjectIn$.subscribe(mySubjectIn$)
  let mw = composed(mySubjectIn$)
  let subsOut = mw.subscribe(subjectOut$)
  return [ subsIn, subsOut ]
}

export class MiddlewareManager {
  private preMiddlewareIn$ = new Subject()
  private preMiddlewareOut$ = new Subject()
  private postMiddlewareIn$ = new Subject()
  private postMiddlewareOut$ = new Subject()

  private preMiddlewares: any[] = []
  private postMiddlewares: any[] = []

  preMiddlewareSubsIn: Subscription
  preMiddlewareSubsOut: Subscription
  postMiddlewareSubsIn: Subscription
  postMiddlewareSubsOut: Subscription

  constructor() {
    this.recomposePre()
    this.recomposePost()
  }

  private recomposePre() {
    [
      this.preMiddlewareSubsIn,
      this.preMiddlewareSubsOut
    ] = recomposeMiddleware(
      this.preMiddlewares,
      this.preMiddlewareIn$,
      this.preMiddlewareOut$,
      this.preMiddlewareSubsIn,
      this.preMiddlewareSubsOut
    )
  }

  private recomposePost() {
    [
      this.postMiddlewareSubsIn,
      this.postMiddlewareSubsOut
    ] = recomposeMiddleware(
      this.postMiddlewares,
      this.postMiddlewareIn$,
      this.postMiddlewareOut$,
      this.postMiddlewareSubsIn,
      this.postMiddlewareSubsOut
    )
  }

  applyPreMiddleware(
    preMiddleware: ((action: Observable<AppAction<any>> ) =>
        Observable<AppAction<any>>)[]
  )  {
    this.preMiddlewares.push(...preMiddleware)
    this.recomposePre()
  }

  applyPostMiddleware(
    postMiddleware: ((state: Observable<AppState>) => Observable<AppState>)[]
  )  {
    this.postMiddlewares.push(...postMiddleware)
    this.recomposePost()
  }

  preMiddleware( action: Observable<AppAction<any>> ):
      Observable<AppAction<any>> {
    action.subscribe(this.preMiddlewareIn$)
    return this.preMiddlewareOut$
  }

  postMiddleware( state: Observable<AppState> ): Observable<AppState> {
    state.subscribe(this.postMiddlewareIn$)
    return this.postMiddlewareOut$
  }

}

```

The plnkr is the result of me trying to extract just enough of what I have to demonstrate that middleware business.

This _does_ work -- that is, in the case of the plnkr, the middleware which is designed to disable the DECREMENT action does do so. 

So, this is what I meant by lazy loading after bootstrap.

~~However, I think I'm leaving live observables in place right now when I reconstitute the middleware, since all I'm doing is unsubscribing from prior middleware when I re-compose the fresh middleware. 
I believe I need to figure out how to disconnect previous version of middleware from the In$ subject. I'm thinking the answer might lie in creating YASTST (Yet Another Subject To Subscribe To)...~~

(I've updated the plnkr with YASTST.... I'm pretty sure there's a much more graceful way to be doing this that involves not just blindly recomposing the entire array of middleware)

Anyway, I'm not trying to make this into the @laurelnaiad show, but I did want to make sure to demonstrate what I meant by lazy loading middleware, even if with a still poor implementation of it.
